lib/connect2.lua

```diff
@@ -80,77 +100,74 @@ local function connect(mapStateToProps, mapDispatchToProps)
 
 			local storeState = self.store:getState()
 
-			local stateMapper = mapStateToProps
-			local stateValues = mapStateToProps(storeState, self.props)
+			local mapStateToProps = mapStateToPropsOrThunk
+			local mappedStoreState = mapStateToProps(storeState, self.props)
 
-			-- mapStateToProps can return a function instead of a state value.
-			-- In this variant, we keep that value as our 'state mapper' instead
-			-- of the original mapStateToProps. This matches react-redux and
-			-- enables connectors to keep instance-level state.
-			if typeof(stateValues) == "function" then
-				stateMapper = stateValues
-				stateValues = stateValues(storeState, self.props)
+			-- mapStateToPropsOrThunk can return a function instead of a state
+			-- value. In this variant, we keep that value as mapStateToProps
+			-- instead of the original mapStateToProps. This matches react-redux
+			-- and enables connectors to keep instance-level state.
+			if typeof(mappedStoreState) == "function" then
+				mapStateToProps = mappedStoreState
+				mappedStoreState = mapStateToProps(storeState, self.props)
 			end
 
-			if stateValues ~= nil and typeof(stateValues) ~= "table" then
+			if mappedStoreState ~= nil and typeof(mappedStoreState) ~= "table" then
 				local message = formatMessage({
 					"mapStateToProps must either return a table, or return another function that returns a table.",
 					"Instead, it returned %q, which is of type %s.",
 				}, {
-					tostring(stateValues),
-					typeof(stateValues),
+					tostring(mappedStoreState),
+					typeof(mappedStoreState),
 				})
 
 				error(message)
 			end
 
-			local dispatchValues = mapDispatchToProps(function(...)
+			local mappedStoreDispatch = mapDispatchToProps(function(...)
 				return self.store:dispatch(...)
 			end)
 
+			local stateUpdater = makeStateUpdater(self.store, mapStateToProps)
+
+			self.mapStateToProps = mapStateToProps
+
+			-- All of the values that we put into state are intended to be used
+			-- by getDerivedStateFromProps, constructed using makeStateUpdater.
```

Comment:
Technically, the mappedStoreDispatch is only used in the stateUpdater itself, but both of those contexts have no knowledge of `self`.  Maybe could word this a little more clearly